src/SparkConduit.sol

```diff
@@ -0,0 +1,314 @@
+
+        assets[asset].positions[ilk].shares += shares;
+        assets[asset].totalShares           += shares;
+
+        emit Deposit(ilk, asset, source, amount);
+    }
+
+    /// @inheritdoc IAllocatorConduit
+    function withdraw(bytes32 ilk, address asset, uint256 maxAmount)
+        external ilkAuth(ilk) returns (uint256 amount)
+    {
+        uint256 liquidityAvailable
+            = IERC20(asset).balanceOf(pool.getReserveData(asset).aTokenAddress);
+
+        amount = liquidityAvailable < maxAmount ? liquidityAvailable : maxAmount;
+
+        // Normally you should update local state first for re-entrancy,
+        // but we need an update-to-date liquidity index for that
+        amount = pool.withdraw(asset, amount, address(this));
+
+        // Convert asset amount to shares
+        uint256 liquidityIndex = pool.getReserveData(asset).liquidityIndex;
+        uint256 shares         = amount.rayDiv(liquidityIndex);
+
+        uint256 pshares     = assets[asset].positions[ilk].shares;
+        uint256 withdrawals = assets[asset].positions[ilk].pendingWithdrawals;
+
+        if (pshares < shares) {
+            uint256 toReturnShares;
+
+            unchecked {
+                toReturnShares = shares - pshares;
+            }
+
+            uint256 toReturn = toReturnShares.rayMul(liquidityIndex);
+            amount -= toReturn;
+            shares = pshares;
+
+            // Return the excess to the pool as it's other user's deposits
+            // Round against the user that is withdrawing
+            pool.supply(asset, toReturn, address(this), 0);
+        }
+
+        assets[asset].positions[ilk].shares -= shares;
+        assets[asset].totalShares           -= shares;
+
+        if (withdrawals > 0) {
+            if (shares <= withdrawals) {
+                assets[asset].positions[ilk].pendingWithdrawals -= shares;
+                assets[asset].totalPendingWithdrawals           -= shares;
+            } else {
+                assets[asset].positions[ilk].pendingWithdrawals = 0;
+                assets[asset].totalPendingWithdrawals           -= withdrawals;
+            }
+        }
+
+        address destination = RegistryLike(registry).buffers(ilk);
+
+        IERC20(asset).transfer(destination, amount);
+
+        emit Withdraw(ilk, asset, destination, amount);
+    }
+
+    /// @inheritdoc IAllocatorConduit
+    function maxDeposit(bytes32, address asset) public view returns (uint256 maxDeposit_) {
+        // Note: Purposefully ignoring any potental supply cap limits on Spark
+        return assets[asset].enabled ? type(uint256).max : 0;
+    }
+
+    /// @inheritdoc IAllocatorConduit
+    function maxWithdraw(bytes32 ilk, address asset) public view returns (uint256 maxWithdraw_) {
+        DataTypes.ReserveData memory reserveData = pool.getReserveData(asset);
+        maxWithdraw_ = assets[asset].positions[ilk].shares.rayMul(reserveData.liquidityIndex);
```

Comment:
Use getNormalizedIncome here instead

Reply:
Lets make a helper function that uses same logic and static aToken repo

Reply:
done